Protected Embeds: Add CLI command to render all `protected-iframe` shortcodes

vaurdan opened the pull request:
This command will parse and render every single `protected-iframe` shortcode present on the site.

It requires a CSV file, that can be generated on the original site, with all the Protected Embeds.

vaurdan committed:
Protected Embeds: Add CLI command to parse and render all `protected-iframe` shortcodes

vaurdan committed:
Changing spaces to tabs

vaurdan committed:
Changing from double quotes to single quotes, where appliable

vaurdan committed:
Fixing code standards spacing issues

vaurdan committed:
Change condition to Yoda style

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+	 * <file>
+	 *  : The CSV file with protected embeds
+	 *
+	 * @subcommand parse-protected-embeds
```
I think `replace` is likely a better term to use than `parse`.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+	/**
+	 * Parses existent protected embeds from WordPress.com
+	 *
+	 * This will replace all the protected-iframe shortcodes with the correct rendered embed
```
We should add a note here and for `import_protected_embeds` how the CSV can be generated.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+		WP_CLI::line( 'Looking for posts with protected-iframe shortcode.' );
+
+		global $wpdb;
+		$query = $wpdb->get_results( "SELECT ID, post_content FROM $wpdb->posts WHERE `post_content` LIKE '%protected-iframe%'" );
```
Should the search be `%[protected-iframe%` for a bit more specificity? (i.e. add the `[`)

mjangda replied:
Any concerns that this could take a long time on a really large dataset? Should we paginate by id and collect all the results first?

vaurdan replied:
> Should the search be %[protected-iframe% for a bit more specificity? (i.e. add the [)

Good call, it will remove any false positive that way. 

> Any concerns that this could take a long time on a really large dataset? Should we paginate by id and collect all the results first?

I though about it, in my test I used a database with around 60k posts and the query took about 5 seconds. Although it was good enough with my dataset, I can see how can this fail with a larger database. I will start implementing pagination.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+		$affected = 0;
+
+		WP_CLI::line( sprintf( 'Found %d posts.', count( $query ) ) );
+		sleep( 3 );
```
This sleep doesn't seem necessary?

vaurdan replied:
I used it to help me see how many posts were found before the `stdout` starts exploding with lines. I will remove it :)

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+				continue;
+			}
+
+			$protected_embeds[ $data['embed_id'] ] = $data;
```
Any idea how many embeds we can expect on some of the larger sites? Could we end up with memory issues? Not something we need to change right now, but worth keeping in mind.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+
+		foreach ( $query as $post ) {
+			// Post does not have a protected-iframe, skip it
+			if ( strpos( $post->post_content, 'protected-iframe' ) === false ) {
```
Should probably search for `[protected-iframe`

mjangda replied:
Minor: Yoda condition.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+
+		// Add custom protected-iframe handler
+		add_shortcode( 'protected-iframe', function ( $attrs ) use ( $protected_embeds ) {
+			return $protected_embeds[ $attrs['id'] ]['html'];
```
Should add some error checking here. What if the shortcode doesn't have an ID or if the embed doesn't exist for some reason?

vaurdan replied:
I can't believe I forgot about that 🤦‍♂️. Will be fixed.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+			$update = $wpdb->update( $wpdb->posts, array( 'post_content' => stripslashes( $parsed_content ) ), array( 'ID' => $post->ID ) );
+
+			if ( ! $update ) {
+				WP_CLI::error( sprintf( 'Error updating %d. Aborting. Updated %d posts so far.', $post->ID, $affected ) );
```
This could be annoying / blocking. Might be better to flag the error and move on to the next one (and display / log a list of all failed conversions at the end).

mjangda replied:
If we do keep the `error`, we should `fclose` the file first.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+			$affected++;
+
+			if ( 0 === $affected % 100 ) {
+				WP_CLI::line( '100 posts updated. Waiting 1 second...' );
```
Minor: would be nice to flag how many are left. If you want to get fancy, we can show a progress bar instead.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+			}
+		}
+
+		WP_CLI::success( sprintf( 'Done! %d posts updated.', $affected ) );
```
Do we need to clear the cache at the end? Either mentioning it as part of the message here or calling `wp_cache_flush` directly.

mjangda commented on class-wpcom-compat-command.php:
```diff
@@ -76,6 +76,103 @@ function import_protected_embeds( $args ) {
+			WP_CLI::line( " * Found {$post->ID}. Updating post content..." );
+
+			// Update the post using $wpdb, to bypass any filters/options being called on `wp_insert_post`.
+			$update = $wpdb->update( $wpdb->posts, array( 'post_content' => stripslashes( $parsed_content ) ), array( 'ID' => $post->ID ) );
```
Why do we need `stripslashes` here?

mjangda reviewed (commented):
Nice work @vaurdan. I left a couple of notes and questions.

vaurdan commented:
Thanks for the review! I'm going to start addressing these issues.

vaurdan committed:
Changing command from `parse-protected-embeds` to `replace-protected-embeds`

vaurdan committed:
Improving command documentation

vaurdan committed:
Changes the search string to include a bracket '['

vaurdan committed:
Removes unnecessary sleep

vaurdan committed:
Yoda condition

vaurdan committed:
Add additional validation to protected-iframe shortcode replacement

vaurdan committed:
Clear the object cache after the replacement is done

vaurdan committed:
Removes useless 'stripslashes' call

vaurdan committed:
Count and store posts that fails the update, instead of aborting

vaurdan committed:
Add progress information every 100 posts

vaurdan commented:
@mjangda I have addressed most of the issues on your feedback.

Regarding the SQL pagination, after some testing, I noticed that it is slower and more unreliable than fetching all the posts at once. Sometimes queries would fail silently, fooling the script to think that the replacement is done. 

Also, given that I need to get the total number of posts (for progress report), instead of using `SQL_CALC_FOUND_ROWS` and multiple queries for every 100 posts, I believe that one single query returning all the posts will be better. So far, on every occasion this script was used, it was fast and flawless, despite the limitless query. If we start seeing issues with this query, we can obviously iterate and add some pagination. What are your thoughts on that?

emrikol commented:
@vaurdan I think this makes sense.  It's possible that we could run out of memory or hit a killed query, but at least this command should be able to restart where it left off when restarted.

Should we add a dry run, so that we can see if the primary SQL query will die?  Maybe we can test a dry run on a giant db?

joshbetz commented:
In order to avoid running duplicative migration steps many times (every time we refresh content), I'm thinking we might be able to use the [shortcode](https://github.com/Automattic/vip-go-wpcom-compat/blob/master/wpcom-shortcodes.php#L10) when we're pre-launch and run this as cleanup either during the launch or even post-launch.

In that case, we'd have the protected embeds table local to the site. So we could pull embeds directly from that table instead of generating a CSV.

We would probably want an additional script to remove the protected embeds table after we've verified that the script finished successfully.

Thoughts?

Also: What happens if someone edits a post with an iframe/script/etc who doesn't have proper permissions? I would have thought any "unsafe" embeds would get removed.

emrikol commented:
> We would probably want an additional script to remove the protected embeds table after we've verified that the script finished successfully.

Maybe we could add this as a flag to this CLI?  `--remove-table-if-empty`?

> What happens if someone edits a post with an iframe/script/etc who doesn't have proper permissions? I would have thought any "unsafe" embeds would get removed.

I think that's a valid concern.  One idea I've had splashing around in my head was to convert these to a Gutenberg embed block, and offer proper editing that way?  I'm not sure of another good option that would save embeds for users without the proper capability.

joshbetz commented:
> Maybe we could add this as a flag to this CLI?

The main thing I like about having two commands is that we can run the first one, then do some kind of verification (diff the generated HTML maybe?), and then run the table cleanup.

joshbetz commented:
> I think that's a valid concern. One idea I've had splashing around in my head was to convert these to a Gutenberg embed block, and offer proper editing that way? I'm not sure of another good option that would save embeds for users without the proper capability.

I think before we spend too much more time on converting these, we need to come up with a way to ensure old posts aren't going to break if they're updated by someone who doesn't have the `unfiltered_html` cap.

emrikol commented:
> I think before we spend too much more time on converting these, we need to come up with a way to ensure old posts aren't going to break if they're updated by someone who doesn't have the `unfiltered_html` cap.

Could we wrap the reversed embeds in a HTML comment or something, and capture it via the `pre_kses` filter?  Just thinking out loud without testing yet.

joshbetz commented:
Closing this per our call in favor of #23 and #24.
